src/AbstractType.php

```diff
@@ -0,0 +1,29 @@
+<?php declare(strict_types=1);
+
+namespace Fortuneglobe\Types;
+
+use Fortuneglobe\Types\Interfaces\RepresentsScalarValue;
+
+/**
+ * Class AbstractType
+ * @package Fortuneglobe\Types
+ */
+abstract class AbstractType
+{
+	final public function __toString() : string
+	{
+		return $this->toString();
+	}
+
+	abstract public function toString() : string;
+
+	public function jsonSerialize() : string
+	{
+		return $this->toString();
+	}
+
+	public function equals( RepresentsScalarValue $other ) : bool
+	{
+		return ($this->toString() === $other->toString() && static::class === get_class( $other ));
```

Comment:
While `toString()` isn't implemented in the `AbstractType` yet, I assume it's recommended to implement `toString()` in all inherited types separately, depending on how these types need to be represented by string and not just a `(string)$this->value`.

So I would prefer something like this ...

```php
abstract class AbstractType
{
	protected $value;

	/* ... */

	public function equals( RepresentsScalarValue $other ) : bool
	{
		return ($this->value === $other->value && static::class === get_class( $other ));
	}

	/* ... */
}
```

You don't need a field member in every concrete type class, still get a unqiue comparison of types with different string representations you don't have to take care of with slightly but existant performance increase while you don't have any conversion operations by `toString()`.

And in the concrete types you cann still use `$this->value` while it's protected accessible.

Reply:
Your suggestion is not possible here, because the `equals()` method expects an interface. Interfaces do not define members, but only methods. So I can only rely on `toString()` to be defined in `$other`. 

AbstractType defines `toString()` as abstract, so it is defined and must be implemented in extending classes. 

We could also add `implements RepresentsScalarType` to AbstractType, but that would be duplicate that interface, since all extending classes implement an interface that extends `RepresentsScalarType`. That's why I left it out in AbstractType.

Reply:
So why don't we define a non-return-type protected `getValue()` in the interface, implementing this method in the abstract type and use this method call in `equals()` instead? So we can let my mentions as is, except `$value`will be `private` now, and fulfill the requirements of using the interface access only in the comparison.

Reply:
Well the problem I see with an non-typed `getValue()` is that it is non-typed. That would be feasible for the usage in the `equals()` method because of the strict comparison. In userland code it would lead to confusion, because of that undefined return type.

I think we agreed already on the fact, that all types must be able to represent their values as strings, thus we declared the `toString()` to be a method that can "safely be called by contract". 

IMO an un-typed `getValue()` would take the whole type approach ad absurdum.